app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,534 @@
+    fun clear() = synchronized(this) {
+        spatialCache.clear()
+        wayCache.clear()
+        relationCache.clear()
+        wayGeometryCache.clear()
+        relationGeometryCache.clear()
+        wayIdsByNodeIdCache.clear()
+        relationIdsByElementKeyCache.clear()
+    }
+
+    /** Reduces cache size to the given number of non-empty [tiles], and removes all data
+     *  not contained in the remaining tiles.
+     */
+    fun trim(tiles: Int) = synchronized(this) {
+        spatialCache.trim(tiles)
+        trimNonSpatialCaches()
+    }
+
+    private fun trimNonSpatialCaches() {
+        synchronized(this) {
+            // ways and relations with at least one element in cache should not be removed
+            val (wayIds, relationIds) = getWayAndRelationIdsWithElementsInSpatialCache()
+
+            wayCache.keys.retainAll { it in wayIds }
+            relationCache.keys.retainAll { it in relationIds }
+            wayGeometryCache.keys.retainAll { it in wayIds }
+            relationGeometryCache.keys.retainAll { it in relationIds }
+
+            // now clean up wayIdsByNodeIdCache and relationIdsByElementKeyCache
+            wayIdsByNodeIdCache.keys.retainAll { spatialCache.get(it) != null }
+            relationIdsByElementKeyCache.keys.retainAll {
+                when (it.type) {
+                    ElementType.NODE -> spatialCache.get(it.id) != null
+                    ElementType.WAY -> it.id in wayIds
+                    ElementType.RELATION -> it.id in relationIds
+                }
+            }
+        }
+    }
+
+    private fun getWayAndRelationIdsWithElementsInSpatialCache(): Pair<Set<Long>, Set<Long>> = synchronized(this) {
+        val wayIds = HashSet<Long>(wayCache.size)
+        for (way in wayCache.values) {
+            if (way != null && way.nodeIds.any { spatialCache.get(it) != null })
+                wayIds.add(way.id)
+        }
+
+        fun RelationMember.isCachedWayOrNode(): Boolean =
+            type == ElementType.NODE && spatialCache.get(ref) != null
+                || type == ElementType.WAY && ref in wayIds
+        fun RelationMember.hasCachedMembers(): Boolean =
+            type == ElementType.RELATION
+                && relationCache[ref]?.members?.any { it.isCachedWayOrNode() } == true
+
+        val relationIds = HashSet<Long>(relationCache.size)
+        for (relation in relationCache.values) {
+            if (relation != null && relation.members.any { member ->
+                    member.isCachedWayOrNode() || member.hasCachedMembers()
+            })
+                relationIds.add(relation.id)
+        }
+        return wayIds to relationIds
+    }
+
+    private fun <K, V> MutableMap<K, V?>.getOrPutNullable(key: K, defaultValue: () -> V?): V? {
+        val value = get(key)
+        return if (value == null && !containsKey(key)) {
+            val computed = defaultValue()
+            put(key, computed)
+            computed
+        } else
+            value
+    }
```

Comment:
Actually it's not the same: `getOrPut` only checks for `null` and doesn't care about `containsKey`, so it will fetch if the value is null.
https://github.com/JetBrains/kotlin/blob/ea836fd46a1fef07d77c96f9d7e8d7807f793453/libraries/stdlib/src/kotlin/collections/Maps.kt#L356-L365

But the first two `?` in line 522 can indeed be removed.

Reply:
Ah, I see